subsys/net/lib/lwm2m/lwm2m_rw_opaque.c

```diff
@@ -0,0 +1,165 @@
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#define LOG_MODULE_NAME net_lwm2m_opaque
+#define LOG_LEVEL 4
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+
+#include <stdarg.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <ctype.h>
+
+#include "lwm2m_object.h"
+#include "lwm2m_rw_opaque.h"
+#include "lwm2m_engine.h"
+#include "lwm2m_util.h"
+
+static int get_opaque(struct lwm2m_input_context *in, uint8_t *value,
+		      size_t buflen, struct lwm2m_opaque_context *opaque,
+		      bool *last_block)
+{
+	uint16_t in_len;
+
+	if (opaque->remaining == 0) {
+		coap_packet_get_payload(in->in_cpkt, &in_len);
+
+		if (in_len == 0) {
+			return -ENODATA;
+		}
+
+		if (in->block_ctx != NULL) {
+			uint32_t block_num =
+				in->block_ctx->ctx.current /
+				coap_block_size_to_bytes(
+					in->block_ctx->ctx.block_size);
+
+			if (block_num == 0) {
+				opaque->len = in->block_ctx->ctx.total_size;
+			}
+
+			if (opaque->len == 0) {
+				/* No size1 option provided, use current
+				 * payload size. This will reset on next packet
+				 * received.
+				 */
+				opaque->remaining = in_len;
+			} else {
+				opaque->remaining = opaque->len;
+			}
+
+		} else {
+			opaque->len = in_len;
+			opaque->remaining = in_len;
+		}
+	}
+
+	return lwm2m_engine_get_opaque_more(in, value, buflen,
+					    opaque, last_block);
+}
+
+static int put_opaque(struct lwm2m_output_context *out,
+			  struct lwm2m_obj_path *path, char *buf,
+			  size_t buflen)
+{
+	int ret;
+
+	ret = buf_append(CPKT_BUF_WRITE(out->out_cpkt), buf, buflen);
+	if (ret < 0) {
+		return ret;
+	}
+
+	return buflen;
+}
+
+
+const struct lwm2m_writer opaque_writer = {
+	.put_opaque = put_opaque,
+};
+
+const struct lwm2m_reader opaque_reader = {
+	.get_opaque = get_opaque,
+};
+
+int do_read_op_opaque(struct lwm2m_message *msg, int content_format)
+{
+	/* Opaque can only return single resource (instance) */
+	if (msg->path.level < LWM2M_PATH_LEVEL_RESOURCE) {
+		return -EPERM;
+	} else if (msg->path.level > LWM2M_PATH_LEVEL_RESOURCE) {
+		if (!IS_ENABLED(CONFIG_LWM2M_VERSION_1_1)) {
+			return -ENOENT;
+		} else if (msg->path.level > LWM2M_PATH_LEVEL_RESOURCE_INST) {
+			return -ENOENT;
+		}
+	}
+
+	return lwm2m_perform_read_op(msg, content_format);
+}
+
+int do_write_op_opaque(struct lwm2m_message *msg)
+{
+	struct lwm2m_engine_obj_inst *obj_inst = NULL;
+	struct lwm2m_engine_obj_field *obj_field;
+	struct lwm2m_engine_res *res = NULL;
+	struct lwm2m_engine_res_inst *res_inst = NULL;
+	int ret;
+	uint8_t created = 0U;
+
+	ret = lwm2m_get_or_create_engine_obj(msg, &obj_inst, &created);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = lwm2m_engine_validate_write_access(msg, obj_inst, &obj_field);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = lwm2m_engine_get_create_res_inst(&msg->path, &res, &res_inst);
+	if (ret < 0) {
+		return -ENOENT;
+	}
+
+	if (msg->path.level < 3) {
```

Comment:
Not sure if I understand this check anyway, wouldn't `lwm2m_engine_get_create_res_inst()` fail if the path level was lower than `LWM2M_PATH_LEVEL_RESOURCE`?